File: backend/src/main/java/com/taskmanager/service/UserManagementService.java
package com.taskmanager.service;

import com.taskmanager.entity.Role;
import com.taskmanager.entity.User;
import com.taskmanager.repository.UserRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

@Service
@Transactional
public class UserManagementService {

    @Autowired
    private UserRepository userRepository;

    public List<User> getAllUsers() {
        return userRepository.findAll();
    }

    public List<User> getManagers() {
        return userRepository.findByRole(Role.MANAGER);
    }

    public List<User> getUsersAvailableForTaskAssignment() {
        return userRepository.findUsersAvailableForTaskAssignment();
    }

    public void updateUserRole(Long userId, Role newRole) {
        userRepository.updateUserRole(userId, newRole);
    }

    public void promoteUserToManager(Long userId) {
        userRepository.updateUserRole(userId, Role.MANAGER);
    }

    public void demoteManagerToUser(Long userId) {
        userRepository.updateUserRole(userId, Role.USER);
    }

    public Map<String, Object> getUserStatistics() {
        Map<String, Object> statistics = new HashMap<>();
        
        statistics.put("totalUsers", userRepository.countTotalUsers());
        statistics.put("adminUsers", userRepository.countByRole(Role.ADMIN));
        statistics.put("managerUsers", userRepository.countByRole(Role.MANAGER));
        statistics.put("regularUsers", userRepository.countByRole(Role.USER));
        
        List<Object[]> roleStats = userRepository.getUserStatisticsByRole();
        Map<String, Long> roleBreakdown = new HashMap<>();
        for (Object[] stat : roleStats) {
            Role role = (Role) stat[0];
            Long count = (Long) stat[1];
            roleBreakdown.put(role.name(), count);
        }
        statistics.put("roleBreakdown", roleBreakdown);
        
        return statistics;
    }

    public void deleteUser(Long userId) {
        userRepository.deleteById(userId);
    }
}

File: frontend/src/app/services/user-management.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { map } from 'rxjs/operators';

export interface User {
  id: number;
  name: string;
  email: string;
  role: string;
  createdAt: string;
  profilePhotoUrl?: string;
}

export interface UserStatistics {
  totalUsers: number;
  adminUsers: number;
  managerUsers: number;
  regularUsers: number;
  roleBreakdown: {
    [key: string]: number;
  };
}

@Injectable({
  providedIn: 'root'
})
export class UserManagementService {
  private apiUrl = 'http://localhost:8080/api';

  constructor(private http: HttpClient) { }

  // Get all users
  getAllUsers(): Observable<User[]> {
    return this.http.get<User[]>(`${this.apiUrl}/auth/users`);
  }

  // Get all managers
  getAllManagers(): Observable<User[]> {
    return this.http.get<User[]>(`${this.apiUrl}/auth/users`).pipe(
      map(users => users.filter(user => user.role === 'MANAGER'))
    );
  }

  // Get users available for task assignment
  getUsersAvailableForTaskAssignment(): Observable<User[]> {
    return this.http.get<User[]>(`${this.apiUrl}/auth/users`).pipe(
      map(users => users.filter(user => user.role === 'USER'))
    );
  }

  // Update user role
  updateUserRole(userId: number, role: string): Observable<{ message: string }> {
    return this.http.put<{ message: string }>(`${this.apiUrl}/admin/users/${userId}/role`, { role });
  }

  // Promote user to manager
  promoteUserToManager(userId: number): Observable<{ message: string }> {
    return this.http.post<{ message: string }>(`${this.apiUrl}/admin/users/${userId}/promote-to-manager`, {});
  }

  // Demote manager to user
  demoteManagerToUser(userId: number): Observable<{ message: string }> {
    return this.http.post<{ message: string }>(`${this.apiUrl}/admin/users/${userId}/demote-to-user`, {});
  }

  // Get user statistics
  getUserStatistics(): Observable<UserStatistics> {
    return this.http.get<User[]>(`${this.apiUrl}/auth/users`).pipe(
      map(users => {
        const totalUsers = users.length;
        const adminUsers = users.filter(u => u.role === 'ADMIN').length;
        const managerUsers = users.filter(u => u.role === 'MANAGER').length;
        const regularUsers = users.filter(u => u.role === 'USER').length;
        
        const roleBreakdown = {
          ADMIN: adminUsers,
          MANAGER: managerUsers,
          USER: regularUsers
        };
        
        return {
          totalUsers,
          adminUsers,
          managerUsers,
          regularUsers,
          roleBreakdown
        };
      })
    );
  }

  // Delete user
  deleteUser(userId: number): Observable<{ message: string }> {
    return this.http.delete<{ message: string }>(`${this.apiUrl}/admin/users/${userId}`);
  }
}
